Add developer documentation forbidding default parameters in detail APIs (#12978)
rapidsai/cudf: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md

rmm::device_uvector<int32_t> v2{100, s, mr};
```

## Default Parameters

While public libcudf APIs are free to include default function parameters, detail functions should not.
Default memory resource parameters make it easy for developers to accidentally allocate memory using the incorrect resource.
Avoiding default memory resources forces developers to consider each memory allocation carefully.

While streams are not currently exposed in libcudf's API, we plan to do so eventually.
As a result, the same arguments for memory resources also apply to streams.
Public APIs will default to using `cudf::get_default_stream()`.
However, including the same default in detail APIs opens the door for developers to forget to pass in a user-provided stream if one is passed to a public API.
Forcing every API to explicitly pass the stream is intended to prevent such mistakes.

> The memory resources -- and eventually, the stream -- are the final parameters for essentially all public APIs.

Reviewer: [This is just pet-peevery] Github Markdown doesn't automatically render emdashes (and these should be mdash `---` not endash `--`) unless you use the HTML code `&mdash`.  Luckily this works just fine with parentheses.

```suggestion
The memory resource (and eventually the stream) is the final parameter for essentially all public APIs.
```

Author: I wasn't sure about dash rendering either! I left off the third hyphen accidentally, but I agree that a parenthetical is preferable here anyway.

For API consistency, the same is true throughout libcudf's internals.

> Therefore, a consequence of not allowing default streams or mrs is that no parameters in detail APIs may have defaults.

Reviewer: ```suggestion
Therefore, a consequence of not allowing default streams or MRs is that no parameters in detail APIs may have defaults.
```

> Therefore, a consequence of not allowing default streams or mrs is that no parameters in detail APIs may have defaults.

Author: Seems like the consensus line width is 100 columns, with a few violations. I reflowed everything that was off.

> Therefore, a consequence of not allowing default streams or mrs is that no parameters in detail APIs may have defaults.

Reviewer: Yeah, 100, correct. Thanks for reflowing.

using the incorrect resource. Avoiding default memory resources forces developers to consider each
memory allocation carefully.

While streams are not currently exposed in libcudf's API, we plan to do so eventually. As a result,
the same reasons for memory resources also apply to streams. Public APIs default to using
`cudf::get_default_stream()`. However, including the same default in detail APIs opens the door for
developers to forget to pass in a user-provided stream if one is passed to a public API. Forcing
every detail API call to explicitly pass a stream is intended to prevent such mistakes.

The memory resources (and eventually, the stream) are the final parameters for essentially all
public APIs. For API consistency, the same is true throughout libcudf's internals. Therefore, a
consequence of not allowing default streams or MRs is that no parameters in detail APIs may have
defaults.

## NVTX Ranges

In order to aid in performance optimization and debugging, all compute intensive libcudf functions
should have a corresponding NVTX range. libcudf has a convenience macro `CUDF_FUNC_RANGE()` that

> will automatically annotates the lifetime of the enclosing function and uses the function's name as

Reviewer: ```suggestion
automatically annotates the lifetime of the enclosing function and uses the function's name as
```
